### PR TITLE
[GEOS-10154] Feature templating - Add HTML template support

### PR DESCRIPTION
**Pull request by taba90**
[GEOS-10154](https://osgeo-org.atlassian.net/browse/GEOS-10154)

This pr
- add support for html templates to features templating plug-in.
- fix a bug in validation triggered when an xpath function has ../../ and the top level source is not defined in template (since it can be omitted). The bug was making the test in **HTMLComplexFeaturesResponseApiTest** failing.
- allows tests in features-templating-core to run with reuseForks (there was no reason to prevent that in that module)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

File: doc/en/user/source/community/features-templating/directives.rst [hunk 1]
```diff
@@ -76,7 +76,7 @@ The following are the directives available in XML based templates.
      - It has to be the root element of an XML template (:code:`<gft:Template> Template content</gft:Template>`)
    * - defines options to customize the output outside of a feature scope
      - gft:Options
-     - specify it as an element at the beggining of the xml document after the :code:`<gft:Template>` one (:code:`<gft:Options></gft:Options>`). GML options: :code:`<gtf:Namespaces>`,:code:`<gtf:SchemaLocation>`
+     - specify it as an element at the beggining of the xml document after the :code:`<gft:Template>` one (:code:`<gft:Options></gft:Options>`). GML options: :code:`<gtf:Namespaces>`,:code:`<gtf:SchemaLocation>`. HTML options: :code:`<script>`,:code:`<style>`, :code: `<link>`.
```
Comment (bradh) [inline doc/en/user/source/community/features-templating/directives.rst:79]:
```suggestion
     - specify it as an element at the beginning of the xml document after the :code:`<gft:Template>` one (:code:`<gft:Options></gft:Options>`). GML options: :code:`<gtf:Namespaces>`,:code:`<gtf:SchemaLocation>`. HTML options: :code:`<script>`,:code:`<style>`, :code: `<link>`.
```

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/VendorOptions.java [hunk 2]
```diff
@@ -23,6 +23,17 @@
     // vendor option used to customize the separator in GeoJSON flat output attributes
     public static final String SEPARATOR = "separator";
 
+    // vendor option used to define Javascript for XHMTL templates
+    public static final String SCRIPT = "script";
+
+    // vendor option used to define Style for XHMTL templates
```
Comment (bradh) [inline src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/VendorOptions.java:29]:
```suggestion
    // vendor option used to define Style for XHTML templates
```

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java [hunk 3]
```diff
@@ -23,11 +24,10 @@ public String getFormat() {
     }
 
     public static List<SupportedFormat> getByExtension(String extensions) {
```
Comment (bradh) [inline src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java:26]: The argument should perhaps be `extension` (singular)

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java [hunk 4]
```diff
@@ -23,11 +24,10 @@ public String getFormat() {
...
-        if (extensions == null) {
-            return Arrays.asList(SupportedFormat.values());
-        } else if (extensions.equals("xml")) {
-            return Arrays.asList(GML);
-        } else return Arrays.asList(JSONLD, GEOJSON);
+        if (extensions == null) return Arrays.asList(SupportedFormat.values());
+        else if (extensions.equals("xml")) return Arrays.asList(GML);
+        else if (extensions.equals("xhtml")) return Arrays.asList(HTML);
+        else return Arrays.asList(JSONLD, GEOJSON);
```
Comment (bradh) [inline src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java:30]: This code formatting looks really weird.

File: src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLFilteredMappedFeature.xhtml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                </li>
+                <li><span class="caret">Shape</span>
+                    <ul class="nested">
+                        <li>${gsml:shape}</li>
+                    </ul>
+                </li>
+            </ul>
+        </li>
+    </ul>
+</gft:Template>
```
Comment (bradh) [inline src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLFilteredMappedFeature.xhtml:119]:
```suggestion
</gft:Template>

```

File: src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLMappedFeature.xhtml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                </li>
+                <li><span class="caret">Shape</span>
+                    <ul class="nested">
+                        <li>${gsml:shape}</li>
+                    </ul>
+                </li>
+            </ul>
+        </li>
+    </ul>
+</gft:Template>
```
Comment (bradh) [inline src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLMappedFeature.xhtml:116]:
```suggestion
</gft:Template>

```

File: src/community/features-templating/features-templating-wfs/src/test/java/org/geoserver/featurestemplating/response/HTMLGetComplexFeatureResponseTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                doc);
+        assertXpathCount(3, "//html/body/ul/li/ul/li/ul/li/ul/li/ul[./li = '-Py']", doc);
+        assertXpathCount(
+                6, "//html/body/ul/li/ul/li/ul/li/ul/li[./span = 'Composition Parts']", doc);
+        assertXpathCount(6, "//html/body/ul/li/ul/li/ul/li/ul/li/ul/li[./span = 'Part']", doc);
+        assertXpathCount(
+                6, "//html/body/ul/li/ul/li/ul/li/ul/li/ul/li/ul/li[./span = 'Role']", doc);
+        assertXpathCount(
+                5,
+                "//html/body/ul/li/ul/li/ul/li/ul/li/ul/li/ul/li/ul[./li = 'interbedded component']",
```
Comment (bradh) [inline src/community/features-templating/features-templating-wfs/src/test/java/org/geoserver/featurestemplating/response/HTMLGetComplexFeatureResponseTest.java:72]: maybe embedded component? I realise that its just text though.

File: src/community/features-templating/features-templating-ogcapi/src/test/java/org/geoserver/featurestemplating/response/HTMLComplexFeaturesResponseApiTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        assertXpathCount(1, "//html/body/ul/li/ul/li/ul/li/ul/li/ul[./li = 'New Group']", doc);
+        assertXpathCount(1, "//html/body/ul/li/ul/li/ul/li/ul/li/ul[./li = '-Xy']", doc);
+        assertXpathCount(
+                1, "//html/body/ul/li/ul/li/ul/li/ul/li[./span = 'Composition Parts']", doc);
+        assertXpathCount(1, "//html/body/ul/li/ul/li/ul/li/ul/li/ul/li[./span = 'Part']", doc);
+        assertXpathCount(
+                1, "//html/body/ul/li/ul/li/ul/li/ul/li/ul/li/ul/li[./span = 'Role']", doc);
+        assertXpathCount(
+                1,
+                "//html/body/ul/li/ul/li/ul/li/ul/li/ul/li/ul/li/ul[./li = 'interbedded component']",
```
Comment (bradh) [inline src/community/features-templating/features-templating-ogcapi/src/test/java/org/geoserver/featurestemplating/response/HTMLComplexFeaturesResponseApiTest.java:66]: maybe embedded component? I realise that its just text though.

File: src/community/features-templating/features-templating-wfs/src/main/resources/applicationContext.xml [hunk 9]
```diff
@@ -13,5 +13,10 @@
         <constructor-arg ref="geoServer"/>
         <constructor-arg ref="templateConfiguration"/>
     </bean>
+
+    <bean id="HTMLGetFeatureResponse" class="org.geoserver.featurestemplating.wfs.HTMLTemplateResponse">
+        <constructor-arg ref="geoServer"/>
+        <constructor-arg ref="templateConfiguration"/>
+    </bean>
     
 </beans>
```
Comment (bradh) [inline src/community/features-templating/features-templating-wfs/src/main/resources/applicationContext.xml:22]:
```suggestion
</beans>

```

**Review by bradh — CHANGES_REQUESTED** [review]

File: doc/en/user/source/community/features-templating/directives.rst (hunk 1 above)
Comment (taba90) [reply doc/en/user/source/community/features-templating/directives.rst:79]: fixed

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/VendorOptions.java (hunk 2 above)
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/VendorOptions.java]: fixed

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java (hunk 3 above)
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java]: fixed

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java (hunk 4 above)
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/configuration/SupportedFormat.java]: This is how the maven formatting plugin formats it.

File: src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLFilteredMappedFeature.xhtml (hunk 5 above)
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLFilteredMappedFeature.xhtml]: fixed

File: src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLMappedFeature.xhtml (hunk 6 above)
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/test/resources/org/geoserver/featurestemplating/response/HTMLMappedFeature.xhtml]: fixed

File: src/community/features-templating/features-templating-wfs/src/main/resources/applicationContext.xml (hunk 9 above)
Comment (taba90) [reply src/community/features-templating/features-templating-wfs/src/main/resources/applicationContext.xml]: fixed

**Comment by taba90** [comment]
thanks for your review @bradh . Your feedbacks have been applied when possible.

**Review by bradh — APPROVED** [review]

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/writers/XHTMLTemplateWriter.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
+package org.geoserver.featurestemplating.writers;
```
Comment (aaime) [inline src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/writers/XHTMLTemplateWriter.java:1]: Lacks copyright header. Please check the other new files as well to make sure all have one.

**Review by aaime — COMMENTED** [review]
Found that this list in configuration.rst has not been updated:

image

One thing is giving me a bit of pause... having to define the style, head and js blocks as vendor options, instead of just putting them down like a normal HTML document.
On one side it's consistent, because all templates are really just using the "loopable" part of the template and hard-code the begin/end of it, but it also seems a bit of a lost occasion, the head part could have been just used as is I guess. 
Also in other templates the template is still a well formed JSON/GML document, but in the case of XHTML, it's not. Finally, see how Wicket does it, the document is a full HTML one, even if parts of it are ignored.

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/VendorOptions.java [hunk 11]
```diff
@@ -23,6 +23,17 @@
     // vendor option used to customize the separator in GeoJSON flat output attributes
     public static final String SEPARATOR = "separator";
 
+    // vendor option used to define Javascript for XHTML templates
+    public static final String SCRIPT = "script";
+
+    // vendor option used to define Style for XHTML templates
+    public static final String STYLE = "style";
+
+    // vendor option used to define Javascript for XHTML templates
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/VendorOptions.java]: thanks for the catch, the header constant was a leftover I've removed it along with the code using it.

File: src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/writers/XHTMLTemplateWriter.java (hunk 10 above)
Comment (taba90) [reply src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/writers/XHTMLTemplateWriter.java:5]: fixed.

**Comment by taba90** [comment]
@aaime thanks for your review fixes have been applied to documentation and code.
Regarding your points on the implementation of the template:

- Yes ideally it should be possible to define everything in an html like way. The main problem is that due the possibility of a wfs request pointing to more then one layer all the elements referring to pieces outside a feature scope needs to be provided as Options so that in case they need to be merged together it is easier to handle them. Due this to have an immediate way to inform the user that templating mainly applies to features while for elements outside the feature scope there are limitations on what can be templated, there have a been a shift to priviledge an approach where everything outside a feature content templating is provided as "$options".
- The support to have a template more adherent to the final result where ideally every piece of the output can be templated no matter if in the feature scope or outside it will be probably added, but before is needed a clear way to handle the possibility of a multiple featuretype request.
- Wrapping the xhtml format in  a gft:Template element might look weirds, but allows to have a valid xml when defining <gft:Options> in it, and will allow the same also in  a of a more adherent html template (starting with <html>) for which we might still define <gft:Options>.

**Pull request merged**